### PR TITLE
Test and do not add leading slash when there is one already #5473

### PR DESCRIPTION
**Pull request by nicholascus**
Redo url parts concatenations the right way.
Moved forward slash prefixing out of the branch to the main level of the function as that would be more consistent behaviour between cases when config['url'] set and empty... though I would rather remove that slash prefixing altogether L583-L585.

This is redo of https://github.com/Codeception/Codeception/pull/5474

### CODE REVIEW

File: src/Codeception/Module/REST.php [hunk 1]
```diff
@@ -575,11 +575,13 @@ public function sendUNLINK($url, array $linkEntries)
...
-            $url = $this->config['url'] . $url;
-            if ($this->config['url'] && strpos($url, '://') === false && strpos($url, '/') !== 0 && $this->config['url'][0] !== '/') {
-                $url = '/' . $url;
-            }
+        if (strpos($url, '://') === false && $this->config['url']) {
+            $url = $this->config['url']
+             . (substr($this->config['url'], -1) != '/' ? '/' : '')
+             . substr($url, substr($url, 0, 1) == '/' ? 1 : 0);
+        }
+        if (strpos($url, '://') === false && strpos($url, '/') !== 0) {
```
Comment (DavertMik) [inline src/Codeception/Module/REST.php]: Is this line ever reachable?

AFAIK: `$this->config['url']` always has value. 
So `strpos($url, '://') === false && $this->config['url']` should be true for every `strpos($url, '://') === false`, and it will change value of $url so this block won't be executed.

Anyway, could you add tests for REST module so we could be sure about all possible cases handled by this change>

Comment (nicholascus) [reply src/Codeception/Module/REST.php]: I don't know why it was there, I had high level of temptation to just delete it. But thought there was a reason for it and it was not making harm.
The origin of that line is #4796

Comment (DavertMik) [reply src/Codeception/Module/REST.php]: Anyway, could you write a test cases for this in `tests/unit/Codeception/Module/RestTest.php`?

**Comment by nicholascus** [comment]
I put together some test cases. It shows that in this PR url construction works as expected and also follows requirements mentioned in #4796, but current codeception-2.6 would fail some of those test cases.

Test cases:
($configUrl, $requestUrl, $expectedFullUrl)
('v1/', 'healthCheck', '/v1/healthCheck')
('/v1', '/healthCheck', '/v1/healthCheck')
('v1', 'healthCheck', '/v1/healthCheck')  **//this is different in current 2.6**
('http://v1/', '/healthCheck', 'http://v1/healthCheck') **// I really been fixing this**
('http://v1', 'healthCheck', 'http://v1/healthCheck') **//this is different in current 2.6**
('http://v1', 'http://v2/healthCheck', 'http://v2/healthCheck')
('http://v1', '', 'http://v1')
('http://v1', '/', 'http://v1/')
('', 'http://v1', 'http://v1')
('', 'healthCheck', '/healthCheck') **//this is different in current 2.6**
('/', 'healthCheck', '/healthCheck')

`
codeception on 2.6 [!$]
$ php codecept run tests/unit/Codeception/Module/RestTest.php:testRestExecute
Codeception PHP Testing Framework v2.6.0
Powered by PHPUnit 7.5.8 by Sebastian Bergmann and contributors.
Running with seed: 

[Seed] 1748706448

Unit Tests (11) ------------------------
✔ RestTest: Rest execute | #0 (0.11s)
✔ RestTest: Rest execute | #1 (0.00s)
✖ RestTest: Rest execute | #2 (0.01s)
✖ RestTest: Rest execute | #3 (0.01s)
✖ RestTest: Rest execute | #4 (0.00s)
✔ RestTest: Rest execute | #5 (0.00s)
✔ RestTest: Rest execute | #6 (0.00s)
✔ RestTest: Rest execute | #7 (0.00s)
✔ RestTest: Rest execute | #8 (0.00s)
✖ RestTest: Rest execute | #9 (0.00s)
✔ RestTest: Rest execute | #10 (0.01s)
----------------------------------------


Time: 636 ms, Memory: 12.00 MB

There were 4 failures:

---------
1) RestTest: Rest execute | #2
 Test  tests/unit/Codeception/Module/RestTest.php:testRestExecute
Failed asserting that two strings are equal.
- Expected | + Actual
@@ @@
-'/v1/healthCheck'
+'/v1healthCheck'
#1  codeception/tests/unit/Codeception/Module/RestTest.php:469
#2  codeception/src/Codeception/Module/REST.php:605
#3  codeception/src/Codeception/Module/REST.php:467
#4  codeception/tests/unit/Codeception/Module/RestTest.php:482
#5  codeception/codecept:43

---------
2) RestTest: Rest execute | #3
 Test  tests/unit/Codeception/Module/RestTest.php:testRestExecute
Failed asserting that two strings are equal.
- Expected | + Actual
@@ @@
-'http://v1/healthCheck'
+'http://v1//healthCheck'
#1  codeception/tests/unit/Codeception/Module/RestTest.php:469
#2  codeception/src/Codeception/Module/REST.php:605
#3  codeception/src/Codeception/Module/REST.php:467
#4  codeception/tests/unit/Codeception/Module/RestTest.php:482
#5  codeception/codecept:43

---------
3) RestTest: Rest execute | #4
 Test  tests/unit/Codeception/Module/RestTest.php:testRestExecute
Failed asserting that two strings are equal.
- Expected | + Actual
@@ @@
-'http://v1/healthCheck'
+'http://v1healthCheck'
#1  codeception/tests/unit/Codeception/Module/RestTest.php:469
#2  codeception/src/Codeception/Module/REST.php:605
#3  codeception/src/Codeception/Module/REST.php:467
#4  codeception/tests/unit/Codeception/Module/RestTest.php:482
#5  codeception/codecept:43

---------
4) RestTest: Rest execute | #9
 Test  tests/unit/Codeception/Module/RestTest.php:testRestExecute
Failed asserting that two strings are equal.
- Expected | + Actual
@@ @@
-'/healthCheck'
+'healthCheck'
#1  codeception/tests/unit/Codeception/Module/RestTest.php:469
#2  codeception/src/Codeception/Module/REST.php:605
#3  codeception/src/Codeception/Module/REST.php:467
#4  codeception/tests/unit/Codeception/Module/RestTest.php:482
#5  codeception/codecept:43

FAILURES!
Tests: 11, Assertions: 18, Failures: 4.
`

File: src/Codeception/Module/REST.php [hunk 2]
```diff
@@ -575,11 +575,10 @@ public function sendUNLINK($url, array $linkEntries)
...
         // allow full url to be requested
-        if (strpos($url, '://') === false) {
-            $url = $this->config['url'] . $url;
-            if ($this->config['url'] && strpos($url, '://') === false && strpos($url, '/') !== 0 && $this->config['url'][0] !== '/') {
-                $url = '/' . $url;
-            }
+        if (strpos($url, '://') === false && $this->config['url']) {
+            $url = $this->config['url']
+             . (substr($this->config['url'], -1) != '/' && $url ? '/' : '')
+             . substr($url, substr($url, 0, 1) == '/' ? 1 : 0);
```
Comment (Naktibalda) [inline src/Codeception/Module/REST.php]: How about doing this instead?
```php
$url = rtrim($this->config['url'], '/') . '/' . ltrim($url, '/');
```

Comment (nicholascus) [reply src/Codeception/Module/REST.php]: Hi, changed the code as you suggested, looks better and gives minor difference in test case only
`
---------
1) RestTest: Rest execute | #6
Test  tests/unit/Codeception/Module/RestTest.php:testRestExecute
Failed asserting that two strings are equal.
- Expected | + Actual
@@ @@
-'http://v1'
+'http://v1/'
`
which might actually be alright, considering that code is much easier to understand. your thoughts?
shall we make it expected result?

File: tests/unit/Codeception/Module/RestTest.php [hunk 3]
```diff
@@ -442,6 +442,63 @@ public function testAmDigestAuthenticatedThrowsExceptionWithFunctionalModules()
...
+        $module->_inject($connectionModule);
+        $module->_initialize();
+        $module->_before(Stub::makeEmpty('\Codeception\Test\Test'));
+
+        $module->sendGET($requestUrl);
+    }
+
+    public static function configAndRequestUrls()
+    {
+        return array(
```
Comment (Naktibalda) [inline tests/unit/Codeception/Module/RestTest.php]: Please use the short array syntax - `[]`.

File: src/Codeception/Module/REST.php [hunk 4]
```diff
@@ -575,11 +575,8 @@ public function sendUNLINK($url, array $linkEntries)
     protected function execute($method, $url, $parameters = [], $files = [])
     {
         // allow full url to be requested
-        if (strpos($url, '://') === false) {
-            $url = $this->config['url'] . $url;
-            if ($this->config['url'] && strpos($url, '://') === false && strpos($url, '/') !== 0 && $this->config['url'][0] !== '/') {
-                $url = '/' . $url;
-            }
+        if (strpos($url, '://') === false && $this->config['url']) {
+            $url = rtrim($this->config['url'], '/') . ($url && $this->config['url'] ? '/' : '') . ltrim($url, '/');
```
Comment (Naktibalda) [inline src/Codeception/Module/REST.php]: There is no need to check `$this->config['url']` in ternary operator, it was checked in a previous line.
Another edge case to test is when there is a trailing slash in `config['url']` - `array('http://v1/', '', 'http://v1/'),`
removing the slash could be a wrong thing to do, right?

Comment (Naktibalda) [reply src/Codeception/Module/REST.php]: How about handling empty `$url` as a special case?
```php
if ($url === '') {
  $url = $this->config['url'];
} else {
  $url = rtrim($this->config['url'], '/') . '/' . ltrim($url, '/')
}
```

File: tests/unit/Codeception/Module/RestTest.php (hunk 3 above)
Comment (nicholascus) [reply tests/unit/Codeception/Module/RestTest.php]: yep, will do that though it wouldn't change much as 

codeception on bug-5473-rest-module-doubles-leading-slashes [$]
$ grep -RIc ' array(' src/Codeception/ | wc -l
     269

codeception on bug-5473-rest-module-doubles-leading-slashes [$]
$ grep -RIc ' array(' tests/ | wc -l
     646

File: src/Codeception/Module/REST.php (hunk 4 above)
Comment (nicholascus) [reply src/Codeception/Module/REST.php]: sure, done.

File: src/Codeception/Module/REST.php (hunk 1 above)
Comment (nicholascus) [reply src/Codeception/Module/REST.php]: done
